### DISCUSSION TITLE
Remove PDF textcontents without annotations.

### DISCUSSION THREAD

**Discussion by pachhaipurna** [pymupdf/PyMuPDF#3145] [Looking for help]
Is it possible to remove pdf text contents without removing annotations, I am trying to replace pdf page from another pdf.
Currently when I replace pdf page the pdf size is increasing so, I want to clean up the page content at first and then replace with another pdf page. I think this will helps not to increase pdf size.

Currently I am hiding page content with white color and replacing pdf page.

Anyone with this experience, Please help me,
@JorjMcKie Could you please help me ?
Thank you

**Comment by JorjMcKie** [comment]
You are trying to do this with redactions, I suppose?
No, this is not working.

**Reply by JorjMcKie** [reply]
You are talking about text annotation anyway, right? Text highlight annots, rectangles, etc. will not be redacted away by redactions.

**Reply by JorjMcKie** [reply]
And removing text annotations is compliant with the design of redactions - and has legal relevance by the way.

**Comment by JorjMcKie** [comment]
You **_could_** try this hack. Please use it without warranty.
```python
import fitz

doc = fitz.open("input.pdf")
page = doc[0]

# get annotation array
aarray = doc.xref_get_key(page.xref, "Annots")

# set annotation to empty
doc.xref_set_key(page.xref, "Annots", "[]")

# clear all page content
page.add_redact_annot(page.rect)
page.apply_redactions()

# restore old annotation array
doc.xref_set_key(page.xref, "Annots", aarray[1])

# put content of page 'pno' of 'src' PDF onto this page
page.show_pdf_page(page.rect, src, pno)

doc.ez_save("output.pdf")
```
